forge/routes/api/applicationDeviceGroup.js

```diff
@@ -0,0 +1,292 @@
+     * @memberof forge.routes.api.application
+     */
+    app.get('/', {
+        preHandler: app.needsPermission('application:devicegroup:list'),
+        schema: {
+            summary: 'Get a list of device groups in an application',
+            tags: ['Applications'],
+            query: { $ref: 'PaginationParams' },
+            params: {
+                type: 'object',
+                properties: {
+                    applicationId: { type: 'string' }
+                }
+            },
+            response: {
+                200: {
+                    type: 'object',
+                    properties: {
+                        meta: { $ref: 'PaginationMeta' },
+                        count: { type: 'number' },
+                        groups: { type: 'array', items: { $ref: 'DeviceGroupSummary' } }
+                    }
+                },
+                '4xx': {
+                    $ref: 'APIError'
+                }
+            }
+        }
+    }, async (request, reply) => {
+        const paginationOptions = app.db.controllers.Device.getDevicePaginationOptions(request)
+
+        const where = {
+            ApplicationId: request.application.hashid
+        }
+
+        const groupData = await app.db.models.DeviceGroup.getAll(paginationOptions, where)
+        const result = {
+            count: groupData.count,
+            meta: groupData.meta,
+            groups: (groupData.groups || []).map(d => app.db.views.DeviceGroup.deviceGroupSummary(d, { includeApplication: false }))
+        }
+        reply.send(result)
+    })
+
+    /**
+     * Add a new Device Group to an Application
+     * @method POST
+     * @name /api/v1/applications/:applicationId/devicegroups
+     * @memberof forge.routes.api.application
+     */
+    app.post('/', {
+        preHandler: app.needsPermission('application:devicegroup:create'),
+        schema: {
+            summary: 'Add a new Device Group to an Application',
+            tags: ['Applications'],
+            body: {
+                type: 'object',
+                properties: {
+                    name: { type: 'string' },
+                    description: { type: 'string' }
+                },
+                required: ['name']
+            },
+            params: {
+                type: 'object',
+                properties: {
+                    applicationId: { type: 'string' }
+                }
+            },
+            response: {
+                201: {
+                    $ref: 'DeviceGroupSummary'
+                },
+                '4xx': {
+                    $ref: 'APIError'
+                }
+            }
+        }
+    }, async (request, reply) => {
+        const application = request.application
+        const name = request.body.name
+        const description = request.body.description
+
+        const newGroup = await app.db.controllers.DeviceGroup.createDeviceGroup(name, { application, description })

```

Comment:
createDeviceGroup throws `Errors` but there doesn't appear to be exception handling here